### PR TITLE
Implement assert equal for operators

### LINKED ISSUE #5644: Improve operator support for new `assert_equal` function: Group 1
### Context

PennyLane currently has a `qml.equal` function which can be used to check the equality of operators, measurements, and tapes. While this has been extremely useful in testing and validation, it can be difficult to figure out *why* two objects aren't equal.

This is why we are currently have added the framework for an `assert_equal` function:

https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L179

```
>>> from pennylane.ops.functions.equal import assert_equal
>>> assert_equal(qml.S(0), qml.T(0))
AssertionError: op1 and op2 are of different types.  Got <class 'pennylane.ops.qubit.non_parametric_ops.S'> and <class 'pennylane.ops.qubit.non_parametric_ops.T'>.
```

The framework for this function was added in PR #5634 , but the implementation still needs to be updated.

### Implementation Details

Comparison currently relies on a `_equal` single dispatch function, with different "kernels" registered for different types of operators. Currently this function returns `True` or `False`.  We are changing it to return `True`, or a string that explains *why* the two operators are not equal.  This change has already been made for `BasisRotation`:

https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L605

This task is to update `_equal` for

* Generic `Operator` classes: see [`_equal_operators`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L317)
* `Controlled` classes: see [`_equal_controlled`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L386)
* `Pow` classes: see [`_equal_pow`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L418)
* `Adjoint` classes: see [`_equal_adjoint`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L427)
* `Exp` classes: see [`_equal_exp`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L435)

While more single dispatch kernels exist for `_equal`, they will be updated as part of separate tasks.

After this PR, we should be able to do something like:
```
>>> assert_equal(qml.adjoint(qml.RX(0.1, wires=0)), qml.adjoint(qml.RX(0.2, wires=0)))
AssertionError: op1 and op2 have different base operations because op1.data[0] is not equal to op2.data[0]. Got 0.1 and 0.2.
>>> assert_equal(qml.exp(qml.X(0), 0.5j), qml.exp(qml.X(0), 1.0j))
AssertionError: op1 and op2 have different coefficients. Got 0.5j and 1.0j.
```

Use your best judgement as to the contents of the string and we will help refine the message if needed.

### Additional Requirements

To fulfill this task, a PR should:

* Update `_equal` kernels for the above cases
* Update tests in [`test_equal.py`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/functions/test_equal.py). Existing tests like, `test_adjoint_comparison` can be extended, or new tests can be written.
* Add a changelog entry
* Pass black, pylint, and isort checks

### LINKED ISSUE #5648: Improve operator support for new `assert_equal` function: Group 2
### Context

PennyLane currently has a `qml.equal` function which can be used to check the equality of operators, measurements, and tapes. While this has been extremely useful in testing and validation, it can be difficult to figure out *why* two objects aren't equal.

This is why we are currently have added the framework for an `assert_equal` function:

https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L179

```
>>> from pennylane.ops.functions.equal import assert_equal
>>> assert_equal(qml.S(0), qml.T(0))
AssertionError: op1 and op2 are of different types.  Got <class 'pennylane.ops.qubit.non_parametric_ops.S'> and <class 'pennylane.ops.qubit.non_parametric_ops.T'>.
```

The framework for this function was added in PR #5634 , but the implementation still needs to be updated.

### Implementation Details

Comparison currently relies on a `_equal` single dispatch function, with different "kernels" registered for different types of operators. Currently this function returns `True` or `False`.  We are changing it to return `True`, or a string that explains *why* the two operators are not equal.  This change has already been made for `BasisRotation`:

https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L605

This task is to update `_equal` for

* `SProd` classes: see [`_equal_sprod`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L446)
* `ControlledSequence` classes: see [`_equal_controlled_sequence`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L405)
* `Prod` and `Sum` composite classes: [`_equal_prod_and_sum`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L370)
* `Tensor` classes (legacy): [`_equal_tensor`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L460)
* `Hamiltonian` (legacy) clases: [`_equal_hamiltonian`](https://github.com/PennyLaneAI/pennylane/blob/b6e96588f9e3d4236d1f31f73e22566298a22344/pennylane/ops/functions/equal.py#L476)

While more single dispatch kernels exist for `_equal`, they will be updated as part of separate tasks.

After this PR and one from Issue #5644 , we should be able to do something like:
```
>>> assert_equal(2.0 * qml.X(0), 2.0 * qml.X(1))
AssertionError: op1 and op2 have different base operations because op1 in on <Wires = [0]> and op2 is on <Wires = [1]>
>>> assert_equal(qml.X(0) + qml.Y(0), qml.X(0) + qml.Y(0) + qml.Z(0))
AssertionError: op1 and op2 have different lengths.
>>> asset_equal(qml.S(0) @ qml.SX(0), qml.SX(0) @ qml.S(0))
AssertionError: sorted op 0 is different because op1 and op2 are of different types. Got pennylane.ops.qubit.non_parametric_ops.S and pennylane.ops.qubit.non_parametric_ops.SX.
>>> op1 = qml.operation.Tensor(qml.X(0), qml.Y(1))
>>> op2 = qml.operation.Tensor(qml.Y(0), qml.X(1))
>>> assert_equal(op1, op2)
AsssetionError: op1 and op2 have different _obs_data outputs.
```

Use your best judgement as to the contents of the string and we will help refine the message if needed.  The output may recursively depend on `_equal` when comparing component operations.

### Additional Requirements

To fulfill this task, a PR should:

* Update `_equal` kernels for the above cases
* Update tests in [`test_equal.py`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/ops/functions/test_equal.py). Existing tests like, `test_adjoint_comparison` can be extended, or new tests can be written.
* Add a changelog entry
* Pass black, pylint, and isort checks

### PR DESCRIPTION
**Pull request by Tarun-Kumar07**
**Context:**
PennyLane currently has a `qml.equal` function which can be used to check the equality of operators, measurements, and tapes. New framework `assert_equal` has been developed which along with checking equality tells the reason for equality failure. 

**Description of the Change:**
- Implement  support for `assert_equal` for different type of operators mentioned in [#5644](https://github.com/PennyLaneAI/pennylane/issues/5644) and [#5648](https://github.com/PennyLaneAI/pennylane/issues/5648)
- Implementation for the above has been changed such that the negative paths return error messages.
- Existing unit tests have been updated to verify the error message.

**Benefits:**
`assert_equals` for mentioned operators above return more granular reason of failure. This eases debugging of tests.

**Possible Drawbacks:**
More maintenance 

**Related GitHub Issues:**
Fixes #5644 
Fixes #5648

### CODE REVIEW

**Review by albi3ro — APPROVED** [review]
This is looking great! I actually don't have any comments 🎉 

Sorry about taking so long to get to this review...

I'm really looking forward to starting to use this in testing and debugging.  It's going to be so much easier to trouble shoot.

File: pennylane/ops/functions/equal.py [hunk 1]
```diff
@@ -336,112 +354,157 @@ def _equal_operators(
...
+        return f"op1 and op2 have arithmetic depth > 0. Got arithmetic depth {op1.arithmetic_depth}"
+
     if op1.wires != op2.wires:
-        return False
+        return f"op1 and op2 have different wires. Got {op1.wires} and {op2.wires}."
 
     if op1.hyperparameters != op2.hyperparameters:
-        return False
+        return (
+            "The hyperparameter is not equal for op1 and op2.\n"
```
Comment (mudit2812) [inline pennylane/ops/functions/equal.py:370]:
```suggestion
            "The hyperparameters are not equal for op1 and op2.\n"
```

File: pennylane/ops/functions/equal.py [hunk 2]
```diff
@@ -336,112 +354,157 @@ def _equal_operators(
...
+            f"Got {op1.hyperparameters}\n and {op2.hyperparameters}."
+        )
+
+    if not all(
+        qml.math.allclose(d1, d2, rtol=rtol, atol=atol) for d1, d2 in zip(op1.data, op2.data)
+    ):
+        return f"op1 and op2 have different data.\n" f"Got {op1.data} and {op2.data}"
```
Comment (mudit2812) [inline pennylane/ops/functions/equal.py:377]: Surprised this didn't cause a pylint error. Optional suggestion below:

```suggestion
        return f"op1 and op2 have different data.\nGot {op1.data} and {op2.data}"
```

**Review by mudit2812 — COMMENTED** [review]
Looks great! I'll approve once the below comments are addressed and CI continues to pass afterwards 😄

**Review by mudit2812 — APPROVED** [review]
Looks good to me 😄

**Comment by isaacdevlugt** [comment]
Nice work everyone! 🎉 

Just to clarify, @Tarun-Kumar07 and @EuGig how much would both of you say that you contributed to this PR?

**Comment by Tarun-Kumar07** [comment]
Hey @isaacdevlugt , all the changes in this PR were done by me :)

**Comment by Tarun-Kumar07** [comment]
Hi @trbromley,

I noticed that the auto-merge was disabled. Is everything okay?

**Pull request merged**